### DISCUSSION TITLE
Using Oboe, can I create multiple output streams and decide the relationship of input and output streams by myself?

### DISCUSSION THREAD

**Discussion by lilinxiong** [google/oboe#1892] [Q&A]
First of all, I have a low-latency requirement. It is precisely because of this requirement that I decided to use Oboe.

Secondly, our needs are still multi-input and multi-output.

That is to say, the input is not only the audio stream collected in the microphone, but also some extras, such as BGM and so on.

Also like the input stream, the output stream is not only the speaker, but also some other outputs such as: there is another channel to send the codec audio, but this channel may be different from the audio sent to the speaker.

Low-latency Oboe is already satisfied. Can Oboe satisfy the multi-input and multi-output audio scenarios? I saw that DeviceID is required when building Stream, so can it support the one that does not require DeviceID? And it does not make any sound, only produces pcm data, and then does other things, such as: codec.

**Comment by lilinxiong** [comment]
If it can be supported, is there any documentation? Or is there a Demo, looking forward to the reply.

**Comment by philburk** [comment]
Is "BGM" stand for Background Music? Where is that coming from? Are you capturing playback?
Are you decoding and playing the background music?

>  Can Oboe satisfy the multi-input and multi-output audio scenarios?

You cannot have multiple streams reading from the microphone.
Theoretically you can output to the speaker and to a headset at the same time.
We have a plan to implement a test for that: #1845

**Comment by lilinxiong** [comment]
yes, Your understanding of BGM is correct, it is "decoding and playing the background music".
Actually I'm not trying to read multiple streams from the microphone.
What I want to do is a function similar to a dispatcher.

For example: ```AudioInputMicrophoneStream``` has now been added.
```AudioInputBackgroundStream1```.
```AudioInputBackgroundStream2```.

Then there is an ```AudioOutputDispatcherStream```, which is a vector container, which contains ```AudioOutputPlayStream``` and ```AudioOutputEncoderStream```.

Then you can choose according to your needs, mix ```InputMicrophone``` & ```inputBackGround1```, and then give it to ```outputEncoderStream```. Give ```inputBackGround2``` to ```outputPlayStream```.

But after thinking about it, this is different from Oboe's original intention and the problem to be solved. So it is more of a business scenario, a solution using Oboe. Extend the implementation of dispatcher on the basis of low latency.

Of course, if the official considers implementing this solution, then I will be very happy.
@philburk
